Add support for customizing NTP server list

willnewton opened the pull request:
Read a list of NTP servers from config.json if available and configure systemd-timesyncd to use them, otherwise use the defaults. This uses a new service rather than resin-net-config due to boot time ordering issues.

shaunmulligan commented:
@willnewton do you have an example of what a user would have to put into config.json to define new NTP servers?

willnewton commented:
@shaunmulligan it would be a key called ntpServers:

    "ntpServers": "time1.google.com time2.google.com time3.google.com time4.google.com"

I should really document this somewhere but I'm not sure where is appropriate.

agherzan commented:
@willnewton  for now we should have this documented in the README of meta-resin. I'm planning for having some proper documentation with sphinx

willnewton commented:
@agherzan I opened a ticket #741 for adding documentation of the config.json keys.

shaunmulligan commented:
@willnewton this looks good 👍 , we should also have a section in the resin.io docs on `config.json`

agherzan commented on meta-resin-common/recipes-connectivity/resin-ntp-config/resin-ntp-config.bb:
```diff
@@ -0,0 +1,27 @@
+LICENSE = "Apache-2.0"
+LIC_FILES_CHKSUM = "file://${RESIN_COREBASE}/COPYING.Apache-2.0;md5=89aea4e17d99a7cacdbeed46a0096b10"
+
+PR = "r0"
```
can be removed as this is the default value anyway

agherzan commented on meta-resin-common/recipes-support/resin-mounts/resin-mounts/etc-systemd-timesyncd.conf.mount:
```diff
@@ -0,0 +1,15 @@
+Options=bind
+
+[Install]
+WantedBy=systemd-timesyncd.service
```
I would let this be resin-bind like all the other bind mounts.

agherzan commented on meta-resin-common/recipes-connectivity/resin-ntp-config/resin-ntp-config.bb:
```diff
@@ -0,0 +1,27 @@
+inherit allarch systemd
+
+SYSTEMD_SERVICE_${PN} = "resin-ntp-config.service"
+RDEPENDS_${PN} = "bash jq"
```
I don't see any bash specific things in the script so I don't think we should depend on bash.

willnewton replied:
I'll also drop the dep on jq as we don't use it directly.

agherzan replied:
Sounds good.

agherzan commented on meta-resin-common/recipes-connectivity/resin-ntp-config/resin-ntp-config/resin-ntp-config.service:
```diff
@@ -0,0 +1,15 @@
+ExecStart=@BASE_BINDIR@/bash @BINDIR@/resin-ntp-config
+Type=oneshot
+RemainAfterExit=yes
+StandardError=journal+console
```
Any reason to not leave it as default value?

willnewton replied:
Which variable are you referring to? It's a copy and paste from resin-net-config

agherzan replied:
I'm referring to StandardError which by default should be on `inherit`. I'm not sure why we have it in net-config either - most probably by mistake. I remember cleaning some of the service in the past maybe I missed this one unless there is a reason that I can't remember of.

willnewton commented:
I've updated the branch to use a different way of adding wants to timesyncd as discussed with @agherzan . Also added documentation to READE.md as suggested by @shaunmulligan

agherzan commented on meta-resin-common/recipes-support/resin-mounts/resin-mounts.bb:
```diff
@@ -84,5 +90,9 @@ do_install () {
+
+        install -c -m 0644 ${WORKDIR}/systemd-timesyncd.conf ${D}${sysconfdir}/systemd/system/systemd-timesyncd.service.d
+        # Also add to resin-bind wants so resin-state-reset will populate the file
+        ln -sf ${systemd_unitdir}/system/etc-systemd-timesyncd.conf.mount ${D}${sysconfdir}/systemd/system/resin-bind.target.wants
```
This is not needed as it already has `WantedBy=resin-bind.service` which when enabled will create the required symlink.

agherzan commented on meta-resin-common/recipes-support/resin-mounts/resin-mounts.bb:
```diff
@@ -60,6 +63,8 @@ do_install () {
     install -d ${D}${sysconfdir}/systemd/system/resin-bind.target.wants
     install -c -m 0644 ${WORKDIR}/resin-bind.target ${D}${systemd_unitdir}/system/
 
+    # Create dropin dir for timesyncd.conf mount
```
As I see you already have a before attribute in the `resin-ntp-config.service` which makes this drop in useless now.

agherzan commented on README.md:
```diff
@@ -174,3 +174,7 @@ Boolean. Enable or disable persistent logging on this device.
 
 String. The country in which the device is operating. This is used for setting with WiFi regulatory domain.
 
+### ntpServers
```
I would split this in another commit.

willnewton commented:
@agherzan I've updated the branch to take into account your changes

willnewton commented:
@agherzan now rebased with 50% fewer changelog entries.

willnewton commented:
@resin-jenkins retest this please
